jgrapht-core/src/main/java/org/jgrapht/graph/AsSynchronizedGraph.java

```diff
@@ -0,0 +1,1127 @@
+ * returned Set is backed by the underlying graph, but when a traversal over the set is started via
+ * a method such as iterator(), a snapshot of the underlying Set is copied for iteration purposes.
+ * For <code>edgesOf</code>, <code>incomingEdgesOf</code> and <code>outgoingEdgesOf</code> methods,
+ * the returned Set is a unmodifiable copy of the result produced by the underlying Graph. Users can
+ * control whether those copies should be cached; caching may significantly increase memory
+ * requirements. If users decide to cache those copies and the backing graph's changes don't affect
+ * them, those copies will be returned the next time the method is called. If the backing graph's
+ * changes affect them, they will be removed from cache and re-created the next time the method is
+ * called. If users decide to not cache those copies, the graph will create ephemeral copies every
+ * time the method is called. For other methods returning a Set, the Set is just the backing Graph's
+ * return.
+ * </p>
+ *
+ * <p>
+ * The created Graph's hashCode is equal to the backing set's hashCode. And the created Graph is equal
+ * to another Graph if they are the same Graph or the backing Graph is equal to the other Graph.
+ * </p>
+ *
+ * @param <V> the graph vertex type
+ * @param <E> the graph edge type
+ *
+ * @author CHEN Kui
+ * @since Feb 23, 2018
+ */
+public class AsSynchronizedGraph<V, E>
+    extends GraphDelegator<V, E>
+    implements Graph<V, E>, Serializable
+{
+    private static final long serialVersionUID = 5144561442831050752L;
+
+    // Object on which to synchronize
+    private final Object mutex;
+
+    // A set encapsulating backing vertexSet.
+    private transient CopyOnDemandSet<V> allVerticesSet;
+
+    // A set encapsulating backing edgeSet.
+    private transient CopyOnDemandSet<E> allEdgesSet;
+
+    private CacheStrategy<V, E> cacheStrategy;
+
+    /**
+     * Constructor for AsSynchronizedGraph with strategy of not caching the copies for
+     * <code>edgesOf</code>, <code>incomingEdgesOf</code> and <code>outgoingEdgesOf</code> methods.
+     *
+     * @param g the backing graph (the delegate).
+     */
+    public AsSynchronizedGraph(Graph<V, E> g)
+    {
+        super(g);
+        mutex = this;
+        cacheStrategy = new NoCache();
+        allEdgesSet = new CopyOnDemandSet<>(super.edgeSet(), mutex);
+        allVerticesSet = new CopyOnDemandSet<>(super.vertexSet(), mutex);
+    }
+
+    /**
+     * Constructor for AsSynchronizedGraph with specified cache strategy for <code>edgesOf</code>,
+     * <code>incomingEdgesOf</code> and <code>outgoingEdgesOf</code> methods.
+     *
+     * @param g the backing graph (the delegate).
+     * @param cacheEnabled a flag whether to use cache for those methods, if <tt>true</tt>, cache
+     *        will be used for those methods, otherwise cache will not be used.
+     */
+    public AsSynchronizedGraph(Graph<V, E> g, boolean cacheEnabled)
```

Comment:
I wonder if instead of public constructors, we should use static factory methods with descriptive names so that it's much clearer when we're creating an `AsSynchronizedGraph` that has a background cache or not.

Signatures-wise, I'm thinking of something like:

```java
public static <V, E> Graph<V, E> create(Graph<V, E> graph) {
    . . .
}

public static <V, E> Graph<V, E> withCache(Graph<V, E> graph) {
    . . .
}
```

Reply:
I agree that boolean constructor parameters make calling code hard to understand, and it will only get worse as we add additional options to tweak behavior.

We don't use static constructor methods too many places in JGraphT, but we have started introducing the builder pattern in some places.  Or another alternative would be to define a SynchronizedGraphParams class (with named methods for setting various options) and pass an instance of that to the constructor.

Reply:
Good idea. Thank you for your guidance. I will implement it in my next commit.